### ISSUE TITLE
Request: Support the Immutable Entities

### ISSUE THREAD

**Issue by mikependon** [mikependon/RepoDB#465]
Currently, RepoDb does not support an immutable classes when fetching the data from the database. This is true to all fetch operations like Query, QueryAll, BatchQuery and/or even in the ExecuteQuery method.

See the sample code snippets below:

```charp
public class Customer
{
     public Customer(int id, string name, ...)
     {
          Id = id;
          Name = name;
          ...
     }

     public int Id { get; set; }

     public string Name { get; set; }
}
```

**Note:** In F#, by default, the classes are immutable and that the developers need to explicitly specify the `CLIMutableAttribute` to make it work.

More information will soon to be placed here.

**Comment by GeeWee** [comment]
If you want to get some inspiration, I think [this](src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverterFactory.cs) is the PR where it was it was implemented in System.Text.Json. I linked to the file which seems like a good starting point for the implementation, as the PR is rather noisy.

**Comment by isaacabraham** [comment]
@mikependon FYI - this also seems to be required on Insert operations as well.

**Comment by mikependon** [comment]
@isaacabraham , for as long as you are passing the models, the compiler reflects all the properties that are readable. I would assume that immutable classes should be working by default. However, RepoDb insert operation is automatically setting back the PK/Identity value to the model after the operation, it seems to fail it there. In anyway, the PK/Identity value is also returned by the method itself, you can process afterwards.

Btw, which case do you think it will fail?

**Comment by isaacabraham** [comment]
It fails on the `Insert` call with a null reference exception - I suspect trying to find the setter on the properties which do not exist.

**Comment by mikependon** [comment]
@isaacabraham - this is noted. FYI, this user story has been pinned and is next in line, will be a part of the next version beta release, together with the other F# related request #499. I will look at this and make sure it is covered by the Integration Tests before the beta release.

**Comment by mikependon** [comment]
This has been fixed and is now available at [RepoDb v1.12.3](https://www.nuget.org/packages/RepoDb/1.12.3). Closing this incident ticket now.
